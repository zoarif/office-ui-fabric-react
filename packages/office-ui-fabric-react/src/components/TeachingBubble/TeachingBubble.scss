@import '../../common/common';

// Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license. See LICENSE in the project root for license information.

//
// Office UI Fabric
// --------------------------------------------------
// Teaching Bubble styles

@keyframes bounceAnimation {
  0% { transform: matrix3d(0.5, 0, 0, 0, 0, 0.5, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  1.7% { transform: matrix3d(0.658, 0, 0, 0, 0, 0.703, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  2.35% { transform: matrix3d(0.725, 0, 0, 0, 0, 0.8, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  3.4% { transform: matrix3d(0.83, 0, 0, 0, 0, 0.946, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  4.7% { transform: matrix3d(0.942, 0, 0, 0, 0, 1.084, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  5.11% { transform: matrix3d(0.971, 0, 0, 0, 0, 1.113, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  6.81% { transform: matrix3d(1.062, 0, 0, 0, 0, 1.166, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  7.06% { transform: matrix3d(1.07, 0, 0, 0, 0, 1.165, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  8.76% { transform: matrix3d(1.104, 0, 0, 0, 0, 1.12, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  9.36% { transform: matrix3d(1.106, 0, 0, 0, 0, 1.094, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  10.66% { transform: matrix3d(1.098, 0, 0, 0, 0, 1.035, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  12.16% { transform: matrix3d(1.075, 0, 0, 0, 0, 0.98, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  12.61% { transform: matrix3d(1.067, 0, 0, 0, 0, 0.969, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  14.51% { transform: matrix3d(1.031, 0, 0, 0, 0, 0.948, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  14.96% { transform: matrix3d(1.024, 0, 0, 0, 0, 0.949, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  17.77% { transform: matrix3d(0.99, 0, 0, 0, 0, 0.981, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  18.37% { transform: matrix3d(0.986, 0, 0, 0, 0, 0.989, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  20.52% { transform: matrix3d(0.98, 0, 0, 0, 0, 1.011, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  22.22% { transform: matrix3d(0.983, 0, 0, 0, 0, 1.016, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  26.08% { transform: matrix3d(0.996, 0, 0, 0, 0, 1.003, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  29.93% { transform: matrix3d(1.003, 0, 0, 0, 0, 0.995, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  31.63% { transform: matrix3d(1.004, 0, 0, 0, 0, 0.996, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  37.64% { transform: matrix3d(1.001, 0, 0, 0, 0, 1.002, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  42.74% { transform: matrix3d(0.999, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  45.35% { transform: matrix3d(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  49.9% { transform: matrix3d(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  50% { transform: matrix3d(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  52.15% { transform: matrix3d(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  54.3% { transform: matrix3d(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  56.46% { transform: matrix3d(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  58.61% { transform: matrix3d(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  64.16% { transform: matrix3d(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  69.72% { transform: matrix3d(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  80.83% { transform: matrix3d(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  91.99% { transform: matrix3d(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
  100% { transform: matrix3d(1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1); }
}

@keyframes opacityFadeIn {
  0% {
    opacity: 0;
    animation-timing-function: $ms-animation-ease-2;
  }
  26.26% { opacity: 1; }
  100% { opacity: 1; }
}

.root {
  display: block;
  max-width: 364px;
  border: 0;
  box-shadow: none !important;

  animation-name: (bounceAnimation, opacityFadeIn);
  animation-duration: 2000ms;
  animation-timing-function: linear;
  animation-fill-mode: both;
}

.bodyContent {
  padding: 20px;
}

.headerIsLarge:not(:last-child) {
  margin-bottom: 14px;
}

.headerIsSmall:not(:last-child) {
  margin-bottom: 14px;
}

.headline {
  margin: 0;
  color: $ms-color-white;

  .headerIsLarge & {
    @include ms-font-xxl;
    font-weight: $ms-font-weight-light;
  }

  .headerIsCondensed & {
    @include ms-font-m;
    @include ms-margin-right(10px);
    font-weight: $ms-font-weight-semibold;
  }

  .headerIsSmall & {
    @include ms-font-m;
    @include ms-margin-right(10px);
    font-weight: $ms-font-weight-semibold;
  }
}

.body:not(:last-child) {
  margin-bottom: 20px;
}

.subText {
  margin: 0;
  font-size: $ms-font-size-m;
  color: $ms-color-white;
  font-weight: $ms-font-weight-semilight;
}

.root .closeButton {
  position: absolute;
  @include ms-right(0);
  top: 0;
  color: $ms-color-white;
  font-size: $ms-icon-size-s;
}

.footer {
  :global(.ms-Button):not(:first-child) {
    @include margin-left(20px);
  }
}

// Callout Overrides
.root {
  :global(.ms-Callout-main),
  :global(.ms-Callout-beak),
  :global(.ms-Callout-smallbeak) {
    background: $ms-color-themePrimary;
  }
}

.wideCallout {
  max-width: 456px;

  :global(.ms-TeachingBubble-content) {
    display: flex;
  }

  :global(.ms-TeachingBubble-image) {
    display: flex;
    align-items: center;
    padding-left: 20px;
    max-width: 154px;
  }
}

.wideCallout .bodyContent {
  max-width: 302px;
}

//Primary button style override
.root .primaryButton {
  background-color: $ms-color-white;
  border-color: $ms-color-white;

  :global(.ms-Button-label) {
    @include ms-font-m;
    color: $ms-color-themePrimary;
  }

  &:hover {
    background-color: $ms-color-themeLighter;
    border-color: $ms-color-themeLighter;
  }

  &:focus {
    background-color: $ms-color-themeLighter;
    border-color: $ms-color-white;
  }

  &:active {
    background-color: $ms-color-white;
    border-color: $ms-color-white;
  }
}

//Secondary button style override
.root .secondaryButton {
  background-color: $ms-color-themePrimary;
  border-color: $ms-color-white;

  :global(.ms-Button-label) {
    @include ms-font-m;
    color: $ms-color-white;
  }

  &:hover {
    background-color: $ms-color-themeDarkAlt;
    border-color: $ms-color-white;
  }

  &:focus {
    background-color: $ms-color-themeDarkAlt;
    border-color: $ms-color-white;
  }

  &:active {
    background-color: $ms-color-themePrimary;
    border-color: $ms-color-white;
  }
}
