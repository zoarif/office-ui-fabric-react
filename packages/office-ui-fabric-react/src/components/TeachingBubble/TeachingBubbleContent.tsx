import * as React from 'react';
import { BaseComponent, css } from '../../Utilities';
import { ITeachingBubbleProps } from './TeachingBubble.types';
import { ITeachingBubbleState } from './TeachingBubble';
import { PrimaryButton, DefaultButton, IconButton } from '../../Button';
import { Image, ImageFit } from '../../Image';
import * as stylesImport from './TeachingBubble.scss';
const styles: any = stylesImport;

export class TeachingBubbleContent extends BaseComponent<ITeachingBubbleProps, ITeachingBubbleState> {

  // Specify default props values
  public static defaultProps = {
    hasCondensedHeadline: false,
    imageProps: {
      imageFit: ImageFit.cover,
      width: 364,
      height: 130
    }
  };

  constructor(props: ITeachingBubbleProps) {
    super(props);

    this.state = {
    };
  }

  public render(): JSX.Element {
    const { illustrationImage, primaryButtonProps, secondaryButtonProps, headline, hasCondensedHeadline, hasCloseIcon, onDismiss, closeButtonAriaLabel, hasSmallHeadline } = this.props;

    let imageContent;
    let headerContent;
    let bodyContent;
    let footerContent;
    let closeButton;

    if (illustrationImage && illustrationImage.src) {
      imageContent = (
        <div className={ 'ms-TeachingBubble-header ms-TeachingBubble-image' }>
          <Image { ...illustrationImage as any } />
        </div>
      );
    }

    if (headline) {
      headerContent = (
        <div
          className={ css(
            'ms-TeachingBubble-header',
            hasCondensedHeadline ?
              'ms-TeachingBubble-header--condensed ' + styles.headerIsCondensed
              : hasSmallHeadline
                ? 'ms-TeachingBubble-header--small ' + styles.headerIsSmall
                : 'ms-TeachingBubble-header--large ' + styles.headerIsLarge
          ) }
        >
          <p className={ css('ms-TeachingBubble-headline', styles.headline) } >
            { headline }
          </p>
        </div>
      );
    }

    if (this.props.children) {
      bodyContent = (
        <div className={ css('ms-TeachingBubble-body', styles.body) }>
          <p className={ css('ms-TeachingBubble-subText', styles.subText) }>
            { this.props.children }
          </p>
        </div>
      );
    }

    if (primaryButtonProps || secondaryButtonProps) {
      footerContent = (
        <div className={ css('ms-TeachingBubble-footer', styles.footer) }>
          { primaryButtonProps && (
            <PrimaryButton
              { ...primaryButtonProps }
              className={ css('ms-TeachingBubble-primaryButton', styles.primaryButton, primaryButtonProps.className) }
            />
          ) }
          { secondaryButtonProps && (
            <DefaultButton
              { ...secondaryButtonProps }
              className={ css('ms-TeachingBubble-secondaryButton', styles.secondaryButton, secondaryButtonProps.className) }
            />
          ) }
        </div>
      );
    }

    if (hasCloseIcon) {
      closeButton = (
        <IconButton
          className={ css('ms-TeachingBubble-closebutton', styles.closeButton) }
          iconProps={ { iconName: 'Cancel' } }
          title={ closeButtonAriaLabel }
          ariaLabel={ closeButtonAriaLabel }
          onClick={ onDismiss }
        />
      );
    }

    return (
      <div className={ css('ms-TeachingBubble-content') }>
        { imageContent }
        { closeButton }
        <div className={ css('ms-TeachingBubble-bodycontent', styles.bodyContent) }>
          { headerContent }
          { bodyContent }
          { footerContent }
        </div>
      </div>
    );
  }
}