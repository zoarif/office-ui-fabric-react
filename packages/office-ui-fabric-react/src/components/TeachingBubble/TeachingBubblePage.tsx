import * as React from 'react';
import { LayerHost } from 'office-ui-fabric-react/lib/Layer';
import {
  ExampleCard,
  ComponentPage,
  IComponentDemoPageProps,
  PageMarkdown,
  PropertiesTableSet
} from '@uifabric/example-app-base';
import { TeachingBubbleBasicExample } from './examples/TeachingBubble.Basic.Example';
import { TeachingBubbleCondensedExample } from './examples/TeachingBubble.Condensed.Example';
import { TeachingBubbleIllustrationExample } from './examples/TeachingBubble.Illustration.Example';
import { TeachingBubbleWideIllustrationExample } from './examples/TeachingBubble.WideIllustration.Example';
import { TeachingBubbleSmallHeadlineExample } from './examples/TeachingBubble.SmallHeadline.Example';
import { ComponentStatus } from '../../demo/ComponentStatus/ComponentStatus';
import { TeachingBubbleStatus } from './TeachingBubble.checklist';

const TeachingBubbleBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/examples/TeachingBubble.Basic.Example.tsx') as string;
const TeachingBubbleCondensedExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/examples/TeachingBubble.Condensed.Example.tsx') as string;
const TeachingBubbleIllustrationExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/examples/TeachingBubble.Basic.Example.tsx') as string;
const TeachingBubbleWideIllustrationExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/examples/TeachingBubble.WideIllustration.Example.tsx') as string;
const TeachingBubbleSmallHeadlineExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/examples/TeachingBubble.SmallHeadline.Example.tsx') as string;

export class TeachingBubblePage extends React.Component<IComponentDemoPageProps, any> {
  public render(): JSX.Element {
    return (
      <ComponentPage
        title='TeachingBubble'
        componentName='TeachingBubbleExample'
        componentUrl='https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/TeachingBubble'
        exampleCards={
          <LayerHost>
            <ExampleCard title='TeachingBubble' code={ TeachingBubbleBasicExampleCode }>
              <TeachingBubbleBasicExample />
            </ExampleCard>
            <ExampleCard title='TeachingBubble Condensed' code={ TeachingBubbleCondensedExampleCode }>
              <TeachingBubbleCondensedExample />
            </ExampleCard>
            <ExampleCard title='TeachingBubble with Illustration' code={ TeachingBubbleIllustrationExampleCode }>
              <TeachingBubbleIllustrationExample />
            </ExampleCard>
            <ExampleCard title='TeachingBubble wide with Illustration' code={ TeachingBubbleWideIllustrationExampleCode }>
              <TeachingBubbleWideIllustrationExample />
            </ExampleCard>
            <ExampleCard title='TeachingBubble with small headline' code={ TeachingBubbleSmallHeadlineExampleCode }>
              <TeachingBubbleSmallHeadlineExample />
            </ExampleCard>
          </LayerHost>
        }
        propertiesTables={
          <PropertiesTableSet
            sources={ [
              require<string>('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/TeachingBubble.types.ts')
            ] }
          />
        }
        overview={
          <PageMarkdown>
            { require<string>('!raw-loader!office-ui-fabric-react/src/components/TeachingBubble/docs/TeachingBubbleOverview.md') }
          </PageMarkdown>
        }
        isHeaderVisible={ this.props.isHeaderVisible }
        componentStatus={
          <ComponentStatus
            { ...TeachingBubbleStatus }
          />
        }
      />
    );
  }
}
