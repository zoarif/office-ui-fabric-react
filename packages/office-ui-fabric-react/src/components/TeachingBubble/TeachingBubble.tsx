import * as React from 'react';
import { BaseComponent, css } from '../../Utilities';
import { TeachingBubbleContent } from './TeachingBubbleContent';
import { ITeachingBubbleProps } from './TeachingBubble.types';
import { Callout, ICalloutProps } from '../../Callout';
import { DirectionalHint } from '../../common/DirectionalHint';
import * as stylesImport from './TeachingBubble.scss';
const styles: any = stylesImport;

export interface ITeachingBubbleState {
  isTeachingBubbleVisible?: boolean;
}

export class TeachingBubble extends BaseComponent<ITeachingBubbleProps, ITeachingBubbleState> {
  public static defaultProps = {
    /**
     * Default calloutProps is deprecated in favor of private _defaultCalloutProps.
     * Remove in next release.
     * @deprecated
     */
    calloutProps: {
      beakWidth: 16,
      gapSpace: 0,
      setInitialFocus: true,
      doNotLayer: false,
      directionalHint: DirectionalHint.rightCenter,
    }
  };

  private _defaultCalloutProps: ICalloutProps;

  // Constructor
  constructor(props: ITeachingBubbleProps) {
    super(props);

    this.state = {
    };

    this._defaultCalloutProps = {
      beakWidth: 16,
      gapSpace: 0,
      setInitialFocus: true,
      doNotLayer: false,
      directionalHint: DirectionalHint.rightCenter,
    };
  }

  public render(): JSX.Element {
    const { calloutProps: setCalloutProps, targetElement, onDismiss } = this.props;
    const calloutProps = { ...this._defaultCalloutProps, ...setCalloutProps };

    return (
      <Callout
        target={ targetElement }
        onDismiss={ onDismiss }
        { ...calloutProps }
        className={ css('ms-TeachingBubble', styles.root, this.props.isWide ? styles.wideCallout : null, calloutProps ? calloutProps.className : undefined) }
      >
        <TeachingBubbleContent { ...this.props } />
      </Callout>
    );
  }
}