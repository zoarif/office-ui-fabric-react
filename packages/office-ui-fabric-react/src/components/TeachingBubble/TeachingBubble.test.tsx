import * as React from 'react';

import * as ReactDOM from 'react-dom';
import * as ReactTestUtils from 'react-dom/test-utils';
import * as renderer from 'react-test-renderer';
import { TeachingBubble } from './TeachingBubble';
import { TeachingBubbleContent } from './TeachingBubbleContent';

describe('TeachingBubble', () => {

  it('renders TeachingBubble correctly', () => {
    const component = renderer.create(<TeachingBubble>Content</TeachingBubble>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    const componentContent = renderer.create(<TeachingBubbleContent headline='Title'>Content</TeachingBubbleContent>);
    const treeContent = componentContent.toJSON();
    expect(treeContent).toMatchSnapshot();
  });

  // <Layer> components will lead ReactDOM.findDOMNode(test_component) return null, so the test is based on the teaching bubble content.
  it('renders a label', () => {
    const component = ReactTestUtils.renderIntoDocument(
      <TeachingBubbleContent
        headline='Title'
      />
    );
    const renderedDOM = ReactDOM.findDOMNode(component as React.ReactInstance) as Element;
    const titleElement = renderedDOM.querySelector('.ms-TeachingBubble-headline');

    expect(titleElement!.textContent).toEqual('Title');
  });

  it('merges callout classNames', () => {
    ReactTestUtils.renderIntoDocument<TeachingBubbleContent>(
      <TeachingBubbleContent
        headline='Title'
        calloutProps={ { className: 'foo' } }
      />
    );
    setTimeout(() => {
      const callout = document.querySelector('.ms-Callout') as HTMLElement;
      expect(callout).toBeDefined();
      expect(callout.classList.contains('ms-TeachingBubble')).toBeTruthy();
      expect(callout.classList.contains('foo')).toBeTruthy();
    }, 0);
  });
});
