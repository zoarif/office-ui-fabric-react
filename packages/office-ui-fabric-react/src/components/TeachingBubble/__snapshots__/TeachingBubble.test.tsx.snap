// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeachingBubble renders TeachingBubble correctly 1`] = `
<span
  className="ms-Layer"
/>
`;

exports[`TeachingBubble renders TeachingBubble correctly 2`] = `
<div
  className="ms-TeachingBubble-content"
>
  <div
    className="ms-TeachingBubble-bodycontent"
  >
    <div
      className="ms-TeachingBubble-header ms-TeachingBubble-header--large undefined"
    >
      <p
        className="ms-TeachingBubble-headline"
      >
        Title
      </p>
    </div>
    <div
      className="ms-TeachingBubble-body"
    >
      <p
        className="ms-TeachingBubble-subText"
      >
        Content
      </p>
    </div>
  </div>
</div>
`;
